File: src/controllers/event.py
from datetime import datetime
from src.models.event import Event
from src.models.user import User
from src.models.course import Course
from src.models.university import University
from uuid import uuid4

from src.controllers.controller import Controller
from src.models.profile import Profile
from src.controllers.errors.profile_not_found import ProfileNotFound


class ProfileController(Controller):
    def __init__(self) -> None:
        self.__profiles = []

    def create(self, name: str,start_at: str, end_at: str, description: str, event_picture: str, location: str, is_valid : bool, reward: int, user: User):

        id = str(uuid4())
        event = Event(name, start_at, end_at, description, event_picture, location, is_valid, reward, user)
        self.__profiles.append(event)
        return id

    def find(self, id) -> Profile:
        profile = self.find_by_id(id)
        if profile:
            return profile.as_dict()

        raise ProfileNotFound(f"Profile {id} not found")

    def update(self, id: str, name: str, email: str, sex: str, university: University, profile_picture: str, university_register: str, course: Course, ranking: int, user: User):
        profile = self.find_by_id(id)
        if not profile:
            return
        profile.name = name
        profile.email = email
        profile.sex = sex
        profile.university = university
        profile.profile_picture = profile_picture
        profile.university_register = university_register
        profile.course = course
        profile.ranking = ranking
        user = user
        profile.updated_at = datetime.now()
        return profile.as_dict()

    def delete(self, id: str, user: User) -> str:
        profile = self.find_by_id(id)
        if not profile:
            return
        profile.deleted_by = user
        profile.deleted_at = datetime.now()
        return profile.id

    def find_all(self):
        return [profile.as_dict() for profile in self.__profiles if profile.deleted_at == None]

    def find_by_id(self, id: str) -> Profile:
        for profile in self.__profiles:
            if profile.id == id and profile.deleted_at == None:
                return profile
        raise ProfileNotFound(f"Profile {id} not found")

File: backend/src/views/static/scripts/event.js
function enableEditing(
  nome,
  email,
  sex,
  age,
  university,
  profile_picture,
  university_register,
  course,
  ranking
) {
  document.getElementById(nome).disabled =
    !document.getElementById(nome).disabled;
  document.getElementById(university).disabled =
    !document.getElementById(university).disabled;
  document.getElementById(email).disabled =
    !document.getElementById(email).disabled;
  document.getElementById(sex).disabled =
    !document.getElementById(sex).disabled;
  document.getElementById(age).disabled =
    !document.getElementById(age).disabled;
  document.getElementById(profile_picture).disabled =
    !document.getElementById(profile_picture).disabled;
  document.getElementById(university_register).disabled =
    !document.getElementById(university_register).disabled;
  document.getElementById(course).disabled =
    !document.getElementById(course).disabled;
  document.getElementById(ranking).disabled =
    !document.getElementById(ranking).disabled;
}

async function create() {
  const user = localStorage.getItem("user_id");

  const form = document.querySelector(".create-form");

  const name = form.elements.name.value;
  const start_at = form.elements.start_at.value;
  const end_at = form.elements.end_at.value;
  const description = form.elements.description.value;
  const event_picture = form.elements.event_picture.value;
  const location_ = form.elements.location.value;
  const reward = parseInt(form.elements.reward.value);

  if (new Date(start_at) > new Date(end_at)){
    return alert(`A data inicial tem que ser maior que a final`);
  }

  const response = await fetch(`http://localhost:5000/event/`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "X-On-Behalf-Of": user,
    },
    body: JSON.stringify({
      name,
      start_at,
      end_at,
      description,
      event_picture,
      location: location_,
      reward 
    }),
  });

  if (response.status != 200) {
    body = await response.json();
    return alert(`Algo deu errado - status: ${response.status}, msg: ${body.error}`);
  }

  location.reload();
}

async function save(id) {
  const user = localStorage.getItem("user_id");

  const form = document.querySelector(".form" + id);

  const name = form.elements.name.value;
  const start_at = form.elements.start_at.value;
  const end_at = form.elements.end_at.value;
  const description = form.elements.description.value;
  const event_picture = form.elements.event_picture.value;
  const location_ = form.elements.location.value;
  const reward = parseInt(form.elements.reward.value);

  if (new Date(start_at) > new Date(end_at)){
    return alert(`A data inicial tem que ser maior que a final`);
  }

  const response = await fetch(`http://localhost:5000/event/${id}`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "X-On-Behalf-Of": user,
    },
    body: JSON.stringify({
      name,
      start_at,
      end_at,
      description,
      event_picture,
      location: location_,
      reward 
    }),
  });

  if (response.status != 200) {
    body = await response.json();
    return alert(`Algo deu errado - status: ${response.status}, msg: ${body.error}`);
  }

  location.reload();
}

async function destroy(id) {
  const user = localStorage.getItem("user_id");
  await fetch(`http://localhost:5000/event/${id}`, {
    method: "DELETE",
    headers: {
      "Content-Type": "application/json",
      "X-On-Behalf-Of": user,
    },
  });

  location.reload();
}
